core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v2.go

```diff
@@ -0,0 +1,324 @@
+		}
+		if minUpkeepLogs > capacity {
+			// if there are more logs to fetch than the capacity, fetch the minimum
+			minUpkeepLogs = capacity
+		}
+		logs, remaining := buf.dequeue(start, end, minUpkeepLogs)
+		for _, l := range logs {
+			result = append(result, BufferedLog{ID: buf.id, Log: l})
+			capacity--
+		}
+		remainingLogs += remaining
+	}
+	return result, remainingLogs
+}
+
+func (b *logBuffer) getUpkeepBuffer(uid *big.Int) (*upkeepLogBuffer, bool) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	ub, ok := b.upkeepBuffers[uid.String()]
+	return ub, ok
+}
+
+func (b *logBuffer) setUpkeepBuffer(uid *big.Int, buf *upkeepLogBuffer) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.upkeepBuffers[uid.String()] = buf
+}
+
+type upkeepLogBuffer struct {
+	lggr logger.Logger
+
+	id      *big.Int
+	maxLogs *atomic.Int32
+
+	q       []logpoller.Log
+	visited map[string]int64
+	lock    sync.RWMutex
+}
+
+func newUpkeepLogBuffer(lggr logger.Logger, id *big.Int, maxLogs int) *upkeepLogBuffer {
+	limit := &atomic.Int32{}
+	limit.Add(int32(maxLogs))
+	return &upkeepLogBuffer{
+		lggr:    lggr.With("id", id.String()),
+		id:      id,
+		maxLogs: limit,
+		q:       make([]logpoller.Log, 0, maxLogs),
+		visited: make(map[string]int64),
+	}
+}
+
+func (ub *upkeepLogBuffer) setConfig(maxLogs int) {
+	ub.maxLogs.Store(int32(maxLogs))
+}
+
+func (ub *upkeepLogBuffer) size() int {
+	ub.lock.RLock()
+	defer ub.lock.RUnlock()
+
+	return len(ub.q)
+}
+
+func (ub *upkeepLogBuffer) dequeue(start, end int64, limit int) ([]logpoller.Log, int) {
+	ub.lock.Lock()
+	defer ub.lock.Unlock()
+
+	if len(ub.q) == 0 {
+		return nil, 0
+	}
+
+	var results []logpoller.Log
+	var remaining int
+	updatedLogs := make([]logpoller.Log, 0, ub.maxLogs.Load())
+	for _, l := range ub.q {
+		if l.BlockNumber >= start && l.BlockNumber <= end {
+			if len(results) < limit {
+				results = append(results, l)
+				continue
+			}
+			remaining++
+		}
+		updatedLogs = append(updatedLogs, l)
+	}
+
+	if len(results) > 0 {
+		ub.q = updatedLogs
+	}
+
+	prommetrics.AutomationLogsInLogBuffer.Sub(float64(len(results)))
+
+	return results, remaining
+}
+
+func (ub *upkeepLogBuffer) enqueue(offsetBlock int64, logsToAdd ...logpoller.Log) (int, int) {
+	ub.lock.Lock()
+	defer ub.lock.Unlock()
+
+	logs := ub.q
+	var added int
+	for _, log := range logsToAdd {
+		if log.BlockNumber < offsetBlock {
+			ub.lggr.Debugw("Skipping log from old block", "offsetBlock", offsetBlock, "logBlock", log.BlockNumber)
+			continue
+		}
+		logid := logID(log)
+		if _, ok := ub.visited[logid]; ok {
+			ub.lggr.Debugw("Skipping known log", "offsetBlock", offsetBlock, "logBlock", log.BlockNumber)
+			continue
+		}
+		added++
+		if len(logs) == 0 {
+			logs = append(logs, log)
+		} else {
+			i, _ := sort.Find(len(logs), func(i int) int {
+				return LogComparator(log, logs[i])
+			})
+			if i == len(logs) {
+				logs = append(logs, log)
+			} else {
+				logs = append(logs[:i], append([]logpoller.Log{log}, logs[i:]...)...)
```

Comment:
Would blindly appending the logs and then sorting them once be more efficient?

Reply:
That mostly depends on the number of logs in buffer (let's call it `n`) vs number of logs that are being added (`a`):
1. sort once at the end -> O(n*logn)
2. maintain sorted slice -> O(a*logn)

If we'll think of extreme scenarios then most likely the buffer will contain more logs than what we are adding.

In any case I have both implementation but the first one is commented out, we can benchmark to check what is better and faster